### PR TITLE
Fix S2479 FN: Support utf-8 strings, ignore raw and verbatim strings

### LINKED ISSUE #6291: Fix S2479  FN: Support utf-8 strings
### Description

FN for S2479 (ControlCharacterInString).

Using C#11 utf8 strings, the rule is not raising an issue.

### Repro steps

```csharp
ReadOnlySpan<byte> Utf8NonCompliant = "test"u8; // FN
ReadOnlySpan<byte> Uft8NonCompliant2 = """test"""u8; // FN
```

### Expected behavior

The rule should raise an issue.

### Actual behavior

The rule does not raise an issue.

### PR DESCRIPTION
**Pull request by gregory-paidis-sonarsource**
Fixes #6291

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ControlCharacterInString.CSharp11.cs [hunk 1]
```diff
@@ -15,8 +15,12 @@ void Utf8StringLiterals()
         {
             ReadOnlySpan<byte> Utf8Compliant = "test"u8;
             ReadOnlySpan<byte> Utf8Compliant2 = """test"""u8;
-            ReadOnlySpan<byte> Utf8NonCompliant = "test"u8; // FN
-            ReadOnlySpan<byte> Uft8NonCompliant2 = """test"""u8; // FN
+            ReadOnlySpan<byte> Utf8NonCompliant = "test"u8; // Noncompliant
+            ReadOnlySpan<byte> Uft8NonCompliant2 = """test"""u8; // Noncompliant
+
+            var x = "hello";
+
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ControlCharacterInString.CSharp11.cs:22]: Maybe we could get rid of those empty lines to have a cleaner test

**Review by cristian-ambrosini-sonarsource — APPROVED** [review]
LGTM! Suggested only a small test case clean up

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ControlCharacterInString.CSharp11.cs (hunk 1 above)
Comment (gregory-paidis-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ControlCharacterInString.CSharp11.cs]: you are right, I removed them.

**Comment by gregory-paidis-sonarsource** [comment]
I marked the interpolated cases as Compliant, thus also solving the [#6290](https://github.com/SonarSource/sonar-dotnet/issues/6290) issue.

Right now the rule is only using syntactic analysis.

If we wanted to catch the following case, we would need semantic analysis:

```csharp
const string constant = "test";
string interpolated = $"test{constant}"; // Constant but interpolated string, semantic model is enough
```

Specifically, we would need to call something like `c.SemanticModel.GetConstantValue(c.Node)`, which would be more expensive than what the rule already does.

Futhermore, if we wanted to catch the following case, we would need to use symbolic execution:

```csharp
string variable = coinFlip() ? "test" : "test";
string interpolated = $"test{variable}"; // Non-constant interpolated string, CFG is needed
```

I do not believe that doing either of these upgrades to the rule is valuable, since the user will get the feedback anyways at the original, soon-to-be-interpolated string.

**Review by cristian-ambrosini-sonarsource — APPROVED** [review]

**Comment by csaba-sagi-sonarsource** [comment]
I suggest having an offline talk about the rule, as issues #6290 and #6292 are not correct. Actually raw string literals should not raise issues as they do not have an escape character mechanism. 
For the same reason, the second case in #6291 is not an FN. And also something like
`ReadOnlySpan<byte> Utf8NonCompliant = @"�test"u8; ` should not raise.

**Comment by gregory-paidis-sonarsource** [comment]
> I suggest having an offline talk about the rule, as issues https://github.com/SonarSource/sonar-dotnet/issues/6290 and https://github.com/SonarSource/sonar-dotnet/issues/6292 are not correct.

@csaba-sagi-sonarsource I agree, let's setup a small meeting for these 3 issues.

**Comment by gregory-paidis-sonarsource** [comment]
@csaba-sagi-sonarsource About the issues regarding this rule, since the last time we synced, the changes are:

- Add comments to #6291
- Add comments to #6292
- Simplify/Enrich the Unit Tests

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ControlCharacterInString.cs [hunk 2]
```diff
@@ -34,10 +34,10 @@ public sealed class ControlCharacterInString : SonarDiagnosticAnalyzer
         internal const string DiagnosticId = "S2479";
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ControlCharacterInString.cs:34]:
```suggestion
        private const string DiagnosticId = "S2479";
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ControlCharacterInString.cs [hunk 3]
```diff
@@ -34,10 +34,10 @@ public sealed class ControlCharacterInString : SonarDiagnosticAnalyzer
...
         private const string MessageFormat = "Replace the control character at position {0} by its escape sequence '{1}'.";
 
-        private static readonly DiagnosticDescriptor rule =
+        private static readonly DiagnosticDescriptor Rule =
             DescriptorFactory.Create(DiagnosticId, MessageFormat);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
         private static readonly IDictionary<char, string> EscapedControlCharacters = new Dictionary<char, string>()
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ControlCharacterInString.cs:42]:
```suggestion
        private static readonly IDictionary<char, string> EscapedControlCharacters = new Dictionary<char, string>
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ControlCharacterInString.cs [hunk 4]
```diff
@@ -102,7 +102,9 @@ protected override void Initialize(SonarAnalysisContext context)
         {
             context.RegisterSyntaxNodeActionInNonGenerated(
                 c => CheckControlCharacter(c, ((LiteralExpressionSyntax)c.Node).Token.Text, 0),
-                SyntaxKind.StringLiteralExpression);
+                SyntaxKind.StringLiteralExpression,
+                SyntaxKindEx.Utf8StringLiteralExpression,
+                SyntaxKindEx.Utf8SingleLineRawStringLiteralToken);
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ControlCharacterInString.cs:107]: The registration to the LiteralToken is not needed. Why did you add it?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ControlCharacterInString.CSharp11.cs [hunk 5]
```diff
@@ -4,36 +4,53 @@ namespace Tests.Diagnostics
...
+                nonCompliant
+                }"; // Compliant, verbatim
+
+            var rawWithControlCharacter = $"""test{
+                nonCompliant
+                }"""; // Compliant, raw
+
+            var rawVerbatimWithControlCharacter = @$"""test{
+                nonCompliant
+                }"""; // Compliant, raw+verbatims
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ControlCharacterInString.CSharp11.cs:53]: Not valid raw + verbation does not exist.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Left a last round of comments.

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM! You will need to update the expected issues for the ITs.
